src/api/erc20/Erc20ApiImpl.ts

```diff
@@ -0,0 +1,123 @@
+import { Erc20Api, TxOptionalParams, Receipt, ERC20 } from 'types'
+import BN from 'bn.js'
+import { ZERO } from 'const'
+import { toBN } from 'utils'
+import Erc20ABI from './Erc20ABI'
+
+import Web3 from 'web3'
+
+/**
+ * Basic implementation of Wallet API
+ */
+export class Erc20ApiImpl implements Erc20Api {
+  private _ReferenceERC20: ERC20
+
+  private static _address2ERC20cache: { [K: string]: ERC20 } = {}
+
+  public constructor(web3: Web3) {
+    this._ReferenceERC20 = new web3.eth.Contract(Erc20ABI)
+
+    // TODO remove later
+    ;(window as any).erc20 = this._ReferenceERC20
+  }
+
+  public async balanceOf(tokenAddress: string, userAddress: string): Promise<BN> {
+    console.log('userAddress', userAddress)
+    console.log('tokenAddress', tokenAddress)
+    if (!userAddress || !tokenAddress) return ZERO
+
+    const erc20 = this._getERC20AtAddress(tokenAddress)
+
+    const result = await erc20.methods.balanceOf(userAddress).call()
+    console.log('balance result for token', tokenAddress, 'user', userAddress, '=', result) // this is string
+
+    return toBN(result)
+  }
+
+  public async allowance(tokenAddress: string, userAddress: string, spenderAddress: string): Promise<BN> {
+    if (!userAddress || !tokenAddress) return ZERO
+
+    const erc20 = this._getERC20AtAddress(tokenAddress)
+
+    const result = await erc20.methods.allowance(userAddress, spenderAddress).call()
+    console.log('allowance result', result)
+
+    return toBN(result)
+  }
+
+  public async approve(
+    tokenAddress: string,
+    userAddress: string,
+    spenderAddress: string,
+    amount: BN,
+    txOptionalParams?: TxOptionalParams,
+  ): Promise<Receipt> {
+    const erc20 = this._getERC20AtAddress(tokenAddress)
+
+    const tx = erc20.methods.approve(spenderAddress, amount).send({
+      from: userAddress,
+    })
+
+    if (txOptionalParams && txOptionalParams.onSentTransaction) {
+      tx.once('receipt', txOptionalParams.onSentTransaction)
+    }
+
+    return tx
+  }
+
+  public async transfer(
+    tokenAddress: string,
+    fromAddress: string,
+    toAddress: string,
+    amount: BN,
+    txOptionalParams?: TxOptionalParams,
+  ): Promise<Receipt> {
+    const erc20 = this._getERC20AtAddress(tokenAddress)
+
+    const tx = erc20.methods.transfer(toAddress, amount).send({
+      from: fromAddress,
+    })
+
+    if (txOptionalParams) {
+      tx.once('receipt', txOptionalParams.onSentTransaction)
+    }
+
+    return tx
+  }
+
+  public async transferFrom(
+    senderAddress: string,
+    tokenAddress: string,
+    fromAddress: string,
+    toAddress: string,
+    amount: BN,
+    txOptionalParams?: TxOptionalParams,
+  ): Promise<Receipt> {
+    const erc20 = this._getERC20AtAddress(tokenAddress)
+
+    const tx = erc20.methods.transferFrom(senderAddress, toAddress, amount).send({
+      from: fromAddress,
```

Comment:
But `fromAddress` in `transfer` is the sender address. Unless the sender is from address. And actually, it usually is the same address. Ugh, now I'm confused.

@W3stside @alfetopito @anxolin 
```
erc20.methods.transferFrom(senderAddress, toAddress, amount).send({
      from: fromAddress,
})
// or

erc20.methods.transferFrom(fromAddress, toAddress, amount).send({
      from: senderAddress,
})
```

Reply:
That depends on which pattern we'll stick to.

In my view, the later makes more sense because it reads in English:
 > Transfer from `address` to `address` the `amount`, sent by `address`.
The confusion is because the parameter on `send` is called `from`.

Also, the address the amount will be sent from is not necessarily the address who will be sending the transaction.

Another suggestion, to remove the ambiguity:

```
erc20.methods.transferFrom(sourceAddress, destinationAddress, amount).send({
      from: senderAddress,
})
```